File: app/src/main/java/usb/DeviceHelp.java
package usb;

import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import com.shifen.game.jfcz.utils.CRC16Util;
import java.util.ArrayList;
import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;




public final class DeviceHelp implements SeralToDeviceHelp {


    public OnDataReceiveListener onDataReceiveListener;


    public static byte DELIVER_GOODS = (byte)52;
    public static byte DELIVER_GOODS_RETURN = (byte)53;
    public static byte CHECK_STATE = (byte)54;
    public static byte CHECK_STATE_RETURN = (byte)55;
    public static byte CHECK_GOOD_SNUM = (byte)56;
    public static byte CHECK_GOOD_SNUM_RETURN = (byte)57;
    public static byte RESET = (byte)58;
    public static byte RESET_RETURN = (byte)59;
    public static byte CHECK_TEMPERATURE = (byte)60;
    public static byte CHECK_TEMPERATURE_RETURN = (byte)61;
    public static byte BUFF_START = (byte)129;
    public static byte BUFF_OVER = (byte)250;

    public static final String path = "/dev/ttyS0";
    public static final int baudrate = 9600;

    public static DeviceHelp deviceHelp;


    public SerialHandler serialHandler;

    public static DeviceHelp getInstance() {
        if (deviceHelp == null) {
            synchronized (DeviceHelp.class) {
                if (deviceHelp == null) {
                    deviceHelp = new DeviceHelp();
                }
            }
        }
        return deviceHelp;
    }

    public void initDevice(@Nullable SerialHandler s) {
        SerialPortUtil.getInstance().initSerialPort(path, baudrate, 0);
        SerialPortUtil.getInstance().setOnDataReceiveListener(this);
        serialHandler = s;


    }

    public void setOnDataReceiveListener(OnDataReceiveListener l) {
        this.onDataReceiveListener = l;
    }

    public final void sendData( byte[] bytes) {
        Log.i("ooo", "=======sendData==========");
        Log.i("ooo", this.bytesToHexString(bytes));
        SerialPortUtil.getInstance().sendBuffer(bytes);
    }


    public void onSeralToDeviceHelp(byte[] bytes) {

        Bundle bundle = new Bundle();
        bundle.putByteArray("key", bytes);
        Message message = serialHandler.obtainMessage();
        message.what = 100;
        message.setData(bundle);
        serialHandler.sendMessage(message);
    }

    public final void updateData(byte [] bytes) {

        Log.i("ooo", "=======onDataReceiveBuffListener==========");
        Log.i("ooo", this.bytesToHexString(bytes));
        if(this.onDataReceiveListener != null) {
            onDataReceiveListener.onDataReceive(bytes);
        }


       /* val operateStatusBody = operateStatusBody(number,status)
        val gson = Gson()
        val json = gson.toJson(operateStatusBody)
        val body = RequestBody.create(okhttp3.MediaType.parse("application/json; charset=utf-8"), json)
        ServiceManager.create(OperateService::class.java).operateStatus(body).observeOnMain {}*/
    }


    public final void deliverGoods(int num) {
        Log.i("JFCZApplication", "deliverGoods: " + num);
        byte byte0 = DELIVER_GOODS;
        byte[] byteArrayOf = new byte[]{byte0, (byte)num};
        byte[] bytes = this.crcCheck(byteArrayOf);
        this.sendData(bytes);
    }


    public final void checkState(Integer num) {
        byte byte0 = CHECK_STATE;
        byte byte1 = num == null?0:(byte)num.intValue();
        byte[] byteArrayOf = new byte[]{byte0, byte1};
        byte[] bytes = this.crcCheck(byteArrayOf);
        this.sendData(bytes);
    }

    public final void checkGoodsNum(Integer num) {
        byte byte0 = CHECK_GOOD_SNUM;
        byte byte1 = num == null?0:(byte)num.intValue();
        byte[] byteArrayOf = new byte[]{byte0, byte1};
        byte[] bytes = this.crcCheck(byteArrayOf);
        this.sendData(bytes);
    }

    public final void reset(Integer num) {
        byte byte0 = RESET;
        byte byte1 = num == null?0:(byte)num.intValue();
        byte[] byteArrayOf = new byte[]{byte0, byte1};
        byte[] bytes = this.crcCheck(byteArrayOf);
        this.sendData(bytes);
    }

    public final void checkTemperature() {
        byte byte0 = CHECK_TEMPERATURE;
        int byte1 = 0;
        byte[] byteArrayOf = new byte[]{byte0, (byte)byte1};
        byte[] bytes = this.crcCheck(byteArrayOf);
        this.sendData(bytes);
    }


    public final byte[] crcCheck(byte[] byteArray) {
        byte[] newArr = CRC16Util.appendCrc16(byteArray);
        byte[] bytes = new byte[]{(byte)129, 1, (byte)byteArray.length, newArr[0], newArr[1], newArr[3], newArr[2], (byte)250};
        return bytes;
    }


    public final String bytesToHexString(@Nullable byte[] src) {
        StringBuilder stringBuilder = new StringBuilder("");
        if(src != null && src.length > 0) {
            int i = 0;
            int var4 = src.length - 1;
            if(i <= var4) {
                while(true) {
                    int v = src[i] & 255;
                    String hv = Integer.toHexString(v);
                    stringBuilder.append(" 0x");
                    if(hv.length() < 2) {
                        stringBuilder.append(0);
                        stringBuilder.append(hv);
                    } else {
                        stringBuilder.append(hv);
                    }

                    if(i == var4) {
                        break;
                    }

                    ++i;
                }
            }

            return stringBuilder.toString();
        } else {
            return null;
        }
    }


}

File: app/src/main/java/com/shifen/game/jfcz/usb/DeviceHelp.kt
package com.shifen.game.jfcz.usb

import android.serialport.SerialPortFinder
import android.util.Log
import com.google.gson.Gson
import com.shifen.game.jfcz.model.operateStatusBody
import com.shifen.game.jfcz.services.OperateService
import com.shifen.game.jfcz.services.ServiceManager
import com.shifen.game.jfcz.services.observeOnMain
import com.shifen.game.jfcz.utils.CRC16Util
import okhttp3.RequestBody

/**
 * Created by Administrator on 2018/10/23.
 */
object DeviceHelp{

    fun initDevice() {

        val path = "/dev/ttyS0"
        val baudrate = 9600
        SerialPortUtil.getInstance().initSerialPort(path,baudrate,0)
    }


    fun sendData(bytes: ByteArray) {
        Log.i("JFCZApplication", "=======sendData==========")
        Log.i("JFCZApplication", bytesToHexString(bytes))
        SerialPortUtil.getInstance().sendBuffer(bytes)
    }


    private var BUFF_START:Byte = 0xF0.toByte();
    private var BUFF_PHONE:Byte = 0xF2.toByte();
    private var BUFF_OVER:Byte = 0xF1.toByte();

    var resArr = ArrayList<Byte>();
    fun update(bytes: ByteArray) {
        bytes.forEach {
            resArr.add(it)
        }

        Log.i("JFCZApplication", "resArr = ${resArr.toString()}")
        var tmp1 = resArr[resArr.size - 1];
        if (tmp1 == BUFF_OVER) {
            var last = resArr.lastIndexOf(BUFF_START);
            var newArr = resArr.subList(last + 1, resArr.size - 1);
            var tmp2 = newArr[0];
            if (tmp2 == BUFF_PHONE) {
                updateData(newArr)
            }
            resArr.clear()
        }
    }


    fun updateData(msg: MutableList<Byte>) {
        Log.i("JFCZApplication", "=======onDataReceiveBuffListener==========")
        Log.i("JFCZApplication", bytesToHexString(msg.toByteArray()))

        if (msg[2] == 0x01.toByte() && msg[3] == 53.toByte()) {

            var status = 3
            if (msg[5].toInt() == 1) {
                status= 0
            }
            if (msg[5].toInt() == 0) {
                status= 1
            }
            var number = msg[4].toInt()
            // todo
            number =8
            val operateStatusBody = operateStatusBody(number,status)
            val gson = Gson()
            val json = gson.toJson(operateStatusBody)
            val body = RequestBody.create(okhttp3.MediaType.parse("application/json; charset=utf-8"), json)
            ServiceManager.create(OperateService::class.java).operateStatus(body).observeOnMain {}
        }

    }

    /**
     *@Description: 出货命令
     *@Param: num 柜门编号
     */
    fun deliverGoods(num: Int) {
        Log.i("JFCZApplication","deliverGoods: ${num}")
        // todo
        var num =8
        val byte0 = 52;
        val byte1 = num;
        val byteArrayOf = byteArrayOf(byte0.toByte(), byte1.toByte())
        val bytes =crcCheck(byteArrayOf)
        sendData(bytes)
    }

    /**
     *@Description: 查询柜门锁状态
     *@Param: num 柜门编号，不传查所有
     */
    fun checkState(num: Int?){
        val byte0 = 54;
        val byte1= if (num ==null){ 0x00 }else { num.toByte() }
        val byteArrayOf = byteArrayOf(byte0.toByte(), byte1.toByte())
        val bytes =crcCheck(byteArrayOf)
        sendData(bytes)
    }

    /**
     *@Description: 查询柜门已出货数命令
     *@Param: num 柜门编号，不传查所有
     */
    fun checkGoodsNum(num: Int?){
        val byte0 = 56;
        val byte1= if (num ==null){ 0x00 }else { num.toByte() }
        val byteArrayOf = byteArrayOf(byte0.toByte(), byte1.toByte())
        val bytes =crcCheck(byteArrayOf)
        sendData(bytes)
    }

    /**
     *@Description: 补货复位命令
     *@Param: num 柜门编号，不传查所有
     */
    fun reset(num: Int?){
        val byte0 = 58;
        val byte1= if (num ==null){ 0x00 }else { num.toByte() }
        val byteArrayOf = byteArrayOf(byte0.toByte(), byte1.toByte())
        val bytes =crcCheck(byteArrayOf)
        sendData(bytes)
    }

    /**
     *@Description: 查询温度命令
     */
    fun checkTemperature (){
        var byte0 = 60;
        var byte1 = 0x00
        val byteArrayOf = byteArrayOf(byte0.toByte(), byte1.toByte())
        var bytes =crcCheck(byteArrayOf)
        sendData(bytes)
    }

    /**
     *@Description: crc检查
     *@Param:
     */
    fun crcCheck(byteArray:ByteArray): ByteArray {
        val newArr = CRC16Util.appendCrc16(byteArray);
        val bytes =byteArrayOf(0x81.toByte(), 0x01, byteArray.size.toByte(), newArr[0], newArr[1], newArr[3], newArr[2], 0xFA.toByte())
        return bytes;
    }

    fun bytesToHexString(src: ByteArray?): String? {
        val stringBuilder = StringBuilder("")
        if (src == null || src.size <= 0) {
            return null
        }
        for (i in 0..src.size - 1) {
            val v = src[i].toInt() and 0xFF
            val hv = Integer.toHexString(v)
            stringBuilder.append(" 0x")
            if (hv.length < 2) {
                stringBuilder.append(0)
                stringBuilder.append(hv)
            }else{
                stringBuilder.append(hv)
            }
        }
        return stringBuilder.toString()
    }

}